src/Core/Grid/Filter/CmsPageCategoryFilterFormFactory.php

```diff
@@ -0,0 +1,103 @@
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2018 PrestaShop SA
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Core\Grid\Filter;
+
+use PrestaShop\PrestaShop\Core\Grid\Definition\GridDefinitionInterface;
+use Symfony\Component\Form\Extension\Core\Type\FormType;
+use Symfony\Component\Form\FormInterface;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+
+/**
+ * Class CmsPageCategoryFilterFormFactory is responsible for changing form action to the custom one.
+ */
+final class CmsPageCategoryFilterFormFactory implements GridFilterFormFactoryInterface
+{
+    /**
+     * @var GridFilterFormFactoryInterface
+     */
+    private $formFactory;
+
+    /**
+     * @var UrlGeneratorInterface
+     */
+    private $urlGenerator;
+
+    /**
+     * @var RequestStack
+     */
+    private $requestStack;
+
+    /**
+     * @param GridFilterFormFactoryInterface $formFactory
+     * @param UrlGeneratorInterface $urlGenerator
+     * @param RequestStack $requestStack
+     */
+    public function __construct(
+        GridFilterFormFactoryInterface $formFactory,
+        UrlGeneratorInterface $urlGenerator,
+        RequestStack $requestStack
+    ) {
+        $this->formFactory = $formFactory;
+        $this->urlGenerator = $urlGenerator;
+        $this->requestStack = $requestStack;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function create(GridDefinitionInterface $definition)
+    {
+        $cmsPageCategoryFilterForm = $this->formFactory->create($definition);
+
+        $newCmsPageCategoryFormBuilder = $cmsPageCategoryFilterForm->getConfig()->getFormFactory()->createNamedBuilder(
+            $definition->getId(),
+            FormType::class
+        );
+
+        /** @var FormInterface $categoryFormItem */
+        foreach ($cmsPageCategoryFilterForm as $categoryFormItem) {
+            $newCmsPageCategoryFormBuilder->add(
+                $categoryFormItem->getName(),
+                get_class($categoryFormItem->getConfig()->getType()->getInnerType()),
+                $categoryFormItem->getConfig()->getOptions()
+            );
+        }
+
+        $request = $this->requestStack->getCurrentRequest();
+
+        if (null !== $request) {
+            $newActionUrl = $this->urlGenerator->generate('admin_cms_pages_search_cms_category', [
```

Comment:
So this is how you replace "extra route parameters" with their real value ?

The idea is interesting. I'm just worried that it's not obvious for someone that passing `route_params_extra` parameters is not enough, and that a 2nd class (CmsPageCategoryFilterFormFactory here) is needed to perform this extra processing.

1) Do you think this behavior could be used in other places ? If yes we might define a specific global behavior such as "if you pass parameters with name request_param_cmsCategoryParentId, it will be automatically casted from the request bag"

2) Alternative: if the behavior is not global and requires to use a specific FormFactory, if the FormFactory is not declared we throw an Exception. So we oblige developers to add it.

3) If we dont do 1) or 2), we need at least to explain what's going to happen in the Grid Definition with code comments in the Definition and a php class doc more verbose here: https://github.com/PrestaShop/PrestaShop/pull/10877/files#diff-6d292ee7cbe1fde98b5b802ba4883b7fR36

Reply:
@matks can you explain point 2 in more details  :thinking: 

This whole class is needed just to override grids form and add action attribute to the form in order to keep the id in url  after the form submit: admin-dev/index.php/improve/design/cms-pages/1

This functionality will be needed in all lists which have nested lists inside and want to pass ids like in: admin-dev/index.php/improve/design/cms-pages/1.

If we want to make this form easily reusable maybe we should put the usage in GridDefinition classes ?

Right now  `CmsPageCategoryFilterFormFactory.php` is a [Decorator](https://sourcemaking.com/design_patterns/decorator/php) for form default factory.
@sarjon do you have any suggestions to keep it how it is or introduce something new?

Reply:
For this PR let's keep it like this. Next nested list we handle we will see how to make it generic, with https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Grid/Filter/CategoryFilterFormFactory.php too